### PR TITLE
[LLK] Bind test harness silu/clamp/log/hardtanh/abs to production metal kernels

### LINKED ISSUE #53912: [LLK] tt-llk test harness measures legacy kernels for silu/clamp/log/hardtanh/abs instead of production metal implementations
## Problem

The tt-llk test harness dispatch (`tt_metal/tt-llk/tests/helpers/include/sfpu_operations.h`), shared by the functional suite (`test_sfpu_unary.py`) and the LLK perf pipeline (`perf_eltwise_unary_sfpu.py`), binds several ops to **legacy tt-llk kernels that production does not use on BH/WH**. Production (ttnn → Compute API `*_tile` → metal `llk_sfpu`) uses the metal-layer implementations instead:

| op | harness measures | production uses |
|---|---|---|
| Silu | tt-llk `_calculate_silu_` (legacy piecewise-linear) | metal `calculate_silu` (x·sigmoid via exp_21f) |
| Clamp | tt-llk `_calculate_clamp_` | metal `calculate_clamp` / `calculate_clamp_int32` |
| Log / LogWithBase | tt-llk `_calculate_log_` | metal `calculate_log` |
| Hardtanh | tt-llk `_calculate_hardtanh_` (3× bf16-offset trick) | metal `calculate_hardtanh` (branchless fp32 `sfpi::clamp`) |
| Abs (fp) | tt-llk `_calculate_abs_` | metal `calculate_abs` |

A repo-wide sweep confirms the tt-llk variants have **zero production consumers on BH/WH** (Quasar legitimately uses its tt-llk silu/log — the Quasar harness is a separate file and is not affected). The implementations differ materially (different algorithms/accuracy for silu, hardtanh, log), so for these ops both correctness goldens and perf CI validate/measure dead code, and changes to the production metal kernels are invisible to the LLK pipeline (observed in practice: a verified silu optimization showed byte-identical perf rows).

Related gap: `logsigmoid` is a harness BinaryOp not included in the binary perf sweep, so it has no perf coverage at all.

## Fix

Rebind the five ops in `sfpu_operations.h` to the metal implementations and wire their proper metal inits (`silu_init`/`clamp_init`/`log_init`/`hardtanh_init`/`abs_init`), adapting the clamp/hardtanh parameter encoding (legacy 3× bf16 offsets → fp32 min/max) and keeping `TanhDerivativeLut` as an explicitly-labeled legacy-variant row.

### PR DESCRIPTION
**Pull request by nstamatovicTT**
## Problem
Fixes #53912.

The shared unary-SFPU test dispatch (`tests/helpers/include/sfpu_operations.h`, used by both `test_sfpu_unary.py` and the LLK perf pipeline's `perf_eltwise_unary_sfpu.py`) bound five ops to legacy tt-llk kernels with **zero production consumers on BH/WH** — production (ttnn → Compute API `*_tile`) uses the metal `llk_sfpu` implementations, which differ materially (silu: piecewise-linear vs exp-based; hardtanh: bf16 offset-chain vs fp32 `sfpi::clamp`; log: entirely different polynomial). For these ops, correctness goldens validated dead code and metal-kernel changes were invisible to perf CI.

## Change
- Rebind `SfpuType::{silu, clamp, log, log_with_base, abs}` (and hardtanh's calculate) to the metal kernels, mirroring the compute-API call shapes exactly (template args, fp32-encoded min/max and log-base runtime params).
- Wire the load-bearing metal inits: `log_init` (seeds `vConstFloatPrgm0-2`) and `silu_init` (→ `sigmoid_init<false>`, reciprocal's Prgm0 seed). clamp/hardtanh/abs metal inits reduce to `reset_counters`, so they stay on the bare init (comment updated).
- `log_with_base` golden scale updated to the fp32 encoding (`0x3FB8AA3B` = 1/ln 2), matching `log_with_base_tile`.
- `SfpuType::tanh_derivative_lut` deliberately keeps the tt-llk LUT variant as a labeled legacy row (production sech2 is already swept as `tanh_derivative`).
- Quasar unaffected: separate driver + `sfpu_operations_quasar.h`, and its tt-llk silu/log **are** production there.

## Validation (silicon)
- **BH** (P150): `test_sfpu_unary.py -k '(Silu or Clamp or Hardtanh or Log or Abs) and Float16_b …'` → **203 passed, 0 failed** (43 pre-existing config skips)
- **WH** (N300): same filter → **238 passed, 0 failed** (8 skips)
- Perf smoke (BH): the Silu perf row now measures the metal kernel — MATH_ISOLATE 167,985 cy vs the legacy kernel's 176,305 cy, consistent with metal silu = sigmoid stream (160k) + the `x·σ(x)` multiply.

## Notes for review
- Goldens needed no change beyond the log-base scale: they are torch-based, and `sfpu_clamp` (total-order `min(max(x,lo),hi)`) mirrors the metal max→min chains by construction.
- Known follow-up not in this PR: `logsigmoid` (a harness BinaryOp) is absent from the binary perf sweep entirely.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:nstamatovic/llk-harness-production-bindings)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:nstamatovic/llk-harness-production-bindings)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:nstamatovic/llk-harness-production-bindings)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:nstamatovic/llk-harness-production-bindings)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:nstamatovic/llk-harness-production-bindings)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:nstamatovic/llk-harness-production-bindings)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:nstamatovic/llk-harness-production-bindings)

### CODE REVIEW

File: tt_metal/tt-llk/tests/helpers/include/sfpu_operations.h [hunk 1]
```diff
@@ -933,24 +941,22 @@ void call_unary_sfpu_operation(std::uint32_t dst_index, std::uint32_t math_forma
...
             DST_ACCUM_MODE,
-            _calculate_log_,
-            (APPROX_MODE, true, ITERATIONS),
+            calculate_log,
+            (APPROX_MODE, FAST_MODE, true /* HAS_BASE_SCALING */, is_fp32_dest_acc_en, ITERATIONS),
             dst_index,
             vector_mode,
-            ITERATIONS,
-            0x3DC5u /* 1/ln(2) in fp16a -> log2(x) */);
+            0x3FB8AA3Bu /* 1/ln(2) in fp32 -> log2(x) */);
```
[the remark this answers is not in the document]
Comment (nstamatovicTT) [reply tt_metal/tt-llk/tests/helpers/include/sfpu_operations.h]: Fixed in 223631144cb — the dispatch now passes `IS_BASE_TWO=true` with the `0x3FB8AA3Bu` scale, mirroring the only pairing ttnn emits for that constant (`log_with_base_tile<..., true>`), and the golden switches to `torch.log2`. The exact-log2 arm (`e_float * 2^-23` exponent path) is now the one under test; verified the golden is exact on powers of two.

File: tt_metal/tt-llk/tests/helpers/include/sfpu_operations.h [hunk 2]
```diff
@@ -1197,28 +1203,24 @@ void call_unary_sfpu_operation(std::uint32_t dst_index, std::uint32_t math_forma
...
+            0xBF800000u /* min = -1.0f (fp32) */,
+            0x3F800000u /* max =  1.0f (fp32) */);
     }
     else if constexpr (OPERATION == SfpuType::hardtanh)
     {
         SFPU_UNARY_CALL(
             DST_SYNC_MODE,
             DST_ACCUM_MODE,
-            _calculate_hardtanh_,
+            calculate_hardtanh,
```
[the remark this answers is not in the document]
Comment (nstamatovicTT) [reply tt_metal/tt-llk/tests/helpers/include/sfpu_operations.h]: Checked — clean on both arches. `WITH_COVERAGE` compile-producer of the Hardtanh and Clamp standard + edge profiles: wormhole 34/34 compiled, blackhole 22 compiled with 12 skips that are the pre-existing `dest_acc=No` format constraints (`_skip_bh_unless_fp32` / `_skip_bh_unsupported_float_combo`), not coverage-related — the BH Float32->Float32 `dest_acc=No` variants compiled. The tt-metal#33268 miscompile does not hit `calculate_hardtanh`/`calculate_clamp`, so no `COVERAGE_COMPILE_SKIP_OPS` entry is needed. Result recorded in the commit message of 223631144cb.

File: tt_metal/tt-llk/tests/python_tests/helpers/golden_generators.py [hunk 3]
```diff
@@ -2954,23 +2954,19 @@ def _rdiv(self, x, value=2.0):
...
-        # implement the same function, so sfpu_clamp is a faithful golden here. They agree by
-        # arithmetic rather than by construction, though, so the agreement is pinned in
-        # test_sfpu_domains rather than assumed -- see
-        # test_hardtanh_golden_matches_the_hardtanh_kernel_chain.
+        # Production metal calculate_hardtanh is sfpi::clamp(v, min, max) with fp32 bounds,
+        # i.e. the same SFPU max/min composition sfpu_clamp models, so the golden is faithful
+        # by construction (unlike the legacy tt-llk bf16 offset-chain kernel this harness
+        # bound before, whose agreement was arithmetic -- see
+        # test_hardtanh_golden_matches_the_hardtanh_kernel_chain in test_sfpu_domains,
+        # kept as a regression pin for that equivalence).
```
[the remark this answers is not in the document]
Comment (nstamatovicTT) [reply tt_metal/tt-llk/tests/python_tests/helpers/golden_generators.py]: Fixed in 223631144cb — took the repoint option: the test is now `test_hardtanh_golden_matches_the_clamp_golden`, asserting `_hardtanh == _clamp` on the same probe set (the faithful-by-construction identity), with docstring and assert message rewritten for the metal binding. Also refreshed the two staleness sites this forwards to: the `sfpu_clamp` docstring now describes the shared SFPSWAP max-then-min composition (`calculate_clamp`'s unary_max_min chain / `calculate_hardtanh`'s `sfpi::clamp`), and the Clamp/Hardtanh `SPECIALS_READY_OPS` rationales in `helpers/sfpu_domains.py` name the metal kernels.

File: tt_metal/tt-llk/tests/helpers/include/sfpu_operations.h [hunk 4]
```diff
@@ -106,18 +106,23 @@
 #include "llk_sfpu/ckernel_sfpu_tanhshrink.h"
 #include "llk_sfpu/ckernel_sfpu_trigonometry.h"
 #include "llk_sfpu/ckernel_sfpu_typecast.h"
-#include "sfpu/ckernel_sfpu_abs.h"
+// abs/clamp/hardtanh/log/silu: production (ttnn -> compute API *_tile) uses the
+// metal implementations, NOT the same-named legacy tt-llk kernels, so the
+// harness binds the metal headers for them (issue #53912). The tt-llk
+// tanh_derivative header stays: it backs the explicit legacy-variant
+// SfpuType::tanh_derivative_lut row.
+#include "llk_sfpu/ckernel_sfpu_abs.h"
```
[the remark this answers is not in the document]
Comment (nstamatovicTT) [reply tt_metal/tt-llk/tests/helpers/include/sfpu_operations.h]: Fixed in 223631144cb — duplicate include dropped, and the comment now says abs only switched its dispatch (metal header was already included above; only the legacy include was removed).

File: tt_metal/tt-llk/tests/helpers/include/sfpu_operations.h [hunk 5]
```diff
@@ -1197,28 +1203,24 @@ void call_unary_sfpu_operation(std::uint32_t dst_index, std::uint32_t math_forma
...
+            calculate_hardtanh,
             (APPROX_MODE, ITERATIONS),
             dst_index,
             vector_mode,
-            ITERATIONS,
-            0x3F80u /* p0 = -min = 1.0 (bf16) */,
-            0xC000u /* p1 = -(max-min) = -2.0 (bf16) */,
-            0x3F80u /* p2 = max = 1.0 (bf16) */);
+            0xBF800000u /* min = -1.0f (fp32) */,
+            0x3F800000u /* max =  1.0f (fp32) */);
```
[the remark this answers is not in the document]
Comment (nstamatovicTT) [reply tt_metal/tt-llk/tests/helpers/include/sfpu_operations.h]: Fixed in 223631144cb — the pair is now `CLAMP_MIN_FP32`/`CLAMP_MAX_FP32`, named alongside the other shared dispatch constants at the top of `call_unary_sfpu_operation` with the same moves-with-the-golden note (`sfpu_dispatch_constants.py: CLAMP_MIN/CLAMP_MAX`), and both branches use it.

**Review by ndivnicTT — APPROVED** [review]
LGTM!

**Review by fvranicTT — APPROVED** [review]

**Comment by ndivnicTT** [comment]
@rtawfik01 I remember there was an effort to move all SFPU from LIB to API layer and many were moved. Is it completed, or couple of SFPU-s are still in LIB layer? Should we wrapp up that work (if any) now?
